packages/ERTP/src/mathHelpers/utils.js

```diff
@@ -0,0 +1,66 @@
+import harden from '@agoric/harden';
+import { passStyleOf } from '@agoric/marshal';
+import { assert } from '@agoric/assert';
+
+const makeObjListMathHelpers = (customInsistFn, getIdentifierFn) => {
+  const identity = harden([]);
+  const helpers = harden({
+    doAssertKind: list => {
+      assert(passStyleOf(list) === 'copyArray', 'list must be an array');
+      list.forEach(customInsistFn);
+    },
+    doGetIdentity: _ => identity,
+    doIsIdentity: list =>
+      passStyleOf(list) === 'copyArray' && list.length === 0,
+    doIsGTE: (left, right) => {
+      const leftIds = left.map(getIdentifierFn);
+      const rightIds = right.map(getIdentifierFn);
+      const leftWeakSet = new WeakSet(leftIds);
+      const has = elem => leftWeakSet.has(elem);
+      return rightIds.every(has);
+    },
+    doIsEqual: (left, right) =>
+      helpers.doIsGTE(left, right) && helpers.doIsGTE(right, left),
```

Comment:
A funny observation: if left >= right and left and right have the same number of elements, they must be equal. Twice as fast without more code ;)

Reply:
But there's a hidden assumption here. If `x` and `y` are different elements, but `getIdentifierFn(x) === getIdentifierFn(y)` then this will rule them equal. Maybe that's ok. What's even harder to reason about is if that happens in `doAdd`, the right's element for that identifier with occlude the left's.

We should at least document that this code assumes `getIdentifierFn(x) === getIdentifierFn(y)` iff `sameStructure(x, y)`. We should then check that the use of this by purses and payments obeys that assumption assuming that the mint holder do.

Reply:
> A funny observation: if left >= right and left and right have the same number of elements, they must be equal. Twice as fast without more code ;)

I think this only works if we are not supporting de-duplication (i.e. we assume the user doesn't include duplicates) within the left and right lists (but not necessarily across them). For example, here's a test that failed with this change:

```
    t.ok(doIsEqual(harden([a]), harden([a, a])), `duplication doesn't matter`);
```